tests/gamma/csm_observability_test.py

```diff
@@ -45,39 +92,383 @@ def setUpClass(cls):
+                    "csm_service_name",
+                    "csm_service_namespace_name",
+                    "csm_workload_canonical_service",
+                    "grpc_method",
+                    "grpc_status",
+                    "grpc_target",
+                ]:
+                    self.assertIn(label_key, metric_labels)
+
+            for metric in _HISTOGRAM_SERVER_METRICS:
+                metric_labels = all_results[metric][0].metric.labels
+                for label_key in [
+                    "csm_mesh_id",
+                    "csm_remote_workload_canonical_service",
+                    "csm_remote_workload_cluster_name",
+                    "csm_remote_workload_location",
+                    "csm_remote_workload_name",
+                    "csm_remote_workload_namespace_name",
+                    "csm_remote_workload_project_id",
+                    "csm_remote_workload_type",
+                    "csm_workload_canonical_service",
+                    "grpc_method",
+                    "grpc_status",
+                ]:
+                    self.assertIn(label_key, metric_labels)
+
+            for metric in _COUNTER_CLIENT_METRICS:
+                metric_labels = all_results[metric][0].metric.labels
+                for label_key in [
+                    "grpc_method",
+                    "grpc_target",
+                ]:
+                    self.assertIn(label_key, metric_labels)
+
+            for metric in _COUNTER_SERVER_METRICS:
+                metric_labels = all_results[metric][0].metric.labels
+                for label_key in [
+                    "grpc_method",
+                ]:
+                    self.assertIn(label_key, metric_labels)
+
+        # Testing whether each metric has the right set of monitored resource
+        # label keys
+        with self.subTest("7_check_resource_label_keys"):
+            # all metrics should have the same set of monitored resource labels
+            # which come from the GMP job
+            for metric in _ALL_METRICS:
+                resource_labels = all_results[metric][0].resource.labels
+                for label_key in [
+                    "cluster",
+                    "instance",
+                    "job",
+                    "location",
+                    "namespace",
+                    "project_id",
+                ]:
+                    self.assertIn(label_key, resource_labels)
+
+        # Testing the values of metric labels
+        with self.subTest("8_check_metrics_label_values"):
+            for metric, time_series in all_results.items():
+                series = time_series[0]
+                self.assertEqual(metric, series.metric.type)
+                for label_key, label_value in series.metric.labels.items():
+                    if label_key == "pod" and self._is_client_metric(metric):
+                        # client pod name
+                        self.assertEqual(test_client.hostname, label_value)
```

Comment:
I'm not too happy with how we make all these nested assertion loops with chained if-else. Not very readable. 
I think we should clearly break up server and client metrics in separate dictionaries, and check them separately. 
Then we'll be able to make use of `self.assertContainsSubset()` and `self.assertDictEqual()`.

Reply:
Did a big refactor on this. No more loops checking each label_key and label value separately. Now for each metric, we are checking against a known bag of key: value pair. Some of the values can be `VALUE_NOT_CHECKED`, but for each check, all the values are in a single place now.